### PR TITLE
[jsk_tools] Add test_node_alive feature

### PR DESCRIPTION
**Pull request by YutoUchimi**
## What is this?

This PR add a script which tests if the specified nodes are alive while running `rostest`.

## Why do I need this?

To easily check whether the specified nodes are working correctly during rostest.

## How to use this?

Please check the documentation.
https://github.com/jsk-ros-pkg/jsk_common/pull/1603/files#diff-b05f353a8b35bc8b643654c58bdca7b8

### CODE REVIEW

File: jsk_tools/test/test_node_alive.test [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  <test test-name="test_node_alive"
+        name="test_node_alive"
+        pkg="jsk_tools" type="test_node_alive.py"
+        time-limit="10" retry="1">
+    <rosparam>
+      node_0: /test_node_alive
+      timeout_0: 10
+      node_1: /publish_string
+      timeout_1: 10
```
[the remark this answers is not in the document]
Comment (YutoUchimi) [reply jsk_tools/test/test_node_alive.test]: Thanks for review!
I added test for negative.

**Comment by k-okada** [comment]
Is required tag (https://github.com/jsk-ros-pkg/jsk_pr2eus/commit/54469437cf38a119808b3b266e553e6f7f5ece27) not working for you?

**Comment by YutoUchimi** [comment]
I tested the `required` tag with a simple code.

### test_always_pass.py
```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import unittest

import rospy
import rostest


PKG = 'jsk_tools'
NAME = 'test_always_pass'


class TestAlwaysPass(unittest.TestCase):
    def __init__(self, *args):
        super(TestAlwaysPass, self).__init__(*args)
        rospy.init_node(NAME)

    def test_always_pass(self):
        pass


if __name__ == '__main__':
    rostest.run(PKG, NAME, TestAlwaysPass, sys.argv)
```

### test_always_pass.test
```
<launch>

  <!-- <node name="rqt_2d_plot" -->
  <!--       pkg="jsk_rqt_plugins" type="rqt_2d_plot"/> -->

  <node name="rqt_2d_plot"
        pkg="jsk_rqt_plugins" type="rqt_2d_plot"
        required="true"/>

  <test test-name="test_always_pass"
        name="test_always_pass"
        pkg="jsk_tools" type="test_always_pass.py"
        time-limit="5" retry="0">
  </test>

</launch>
```

### Terminal
```
$ rostest jsk_rqt_plugins test_always_pass.test
... logging to /home/yutouchimi/.ros/log/rostest-libra-31835.log
[ROSUNIT] Outputting test results to /home/yutouchimi/.ros/test_results/jsk_rqt_plugins/rostest-test_test_always_pass.xml
Traceback (most recent call last):
  File "/home/yutouchimi/Projects/mirror_recognition/src/jsk-ros-pkg/jsk_visualization/jsk_rqt_plugins/bin/rqt_2d_plot", line 8, in <module>
    import roslib; roslib.load_manifest('jsk_rqt_plugins')
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslib/launcher.py", line 62, in load_manifest
    sys.path = _generate_python_path(package_name, _rospack) + sys.path
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslib/launcher.py", line 93, in _generate_python_path
    m = rospack.get_manifest(pkg)
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 167, in get_manifest
    return self._load_manifest(name)
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 211, in _load_manifest
    retval = self._manifests[name] = parse_manifest_file(self.get_path(name), self._manifest_name, rospack=self)
  File "/usr/lib/python2.7/dist-packages/rospkg/manifest.py", line 412, in parse_manifest_file
    _static_rosdep_view = init_rospack_interface()
  File "/usr/lib/python2.7/dist-packages/rosdep2/rospack.py", line 60, in init_rospack_interface
    return lookup.get_rosdep_view(DEFAULT_VIEW_KEY)
  File "/usr/lib/python2.7/dist-packages/rosdep2/lookup.py", line 608, in get_rosdep_view
    self._load_view_dependencies(view_key, self.loader)
  File "/usr/lib/python2.7/dist-packages/rosdep2/lookup.py", line 535, in _load_view_dependencies
    loader.load_view(view_key, db, verbose=self.verbose)
  File "/usr/lib/python2.7/dist-packages/rosdep2/rospkg_loader.py", line 94, in load_view
    if view_name not in self.get_loadable_views():
  File "/usr/lib/python2.7/dist-packages/rosdep2/rospkg_loader.py", line 112, in get_loadable_views
    return list(self._rosstack.list()) + [DEFAULT_VIEW_KEY]
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 192, in list
    self._update_location_cache()
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 184, in _update_location_cache
    list_by_path(self._manifest_name, path, cache)
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 61, in list_by_path
    for d, dirs, files in os.walk(path, topdown=True, followlinks=True):
  File "/usr/lib/python2.7/os.py", line 296, in walk
    for x in walk(new_path, topdown, onerror, followlinks):
  File "/usr/lib/python2.7/os.py", line 292, in walk
    yield top, dirs, nondirs
KeyboardInterrupt
[Testcase: testtest_always_pass] ... ok

[ROSTEST]-----------------------------------------------------------------------

[jsk_rqt_plugins.rosunit-test_always_pass/test_always_pass][passed]

SUMMARY
 * RESULT: SUCCESS
 * TESTS: 1
 * ERRORS: 0
 * FAILURES: 0

rostest log file is in /home/yutouchimi/.ros/log/rostest-libra-31835.log
yutouchimi@libra:~/Projects/mirror_recognition/src/jsk-ros-pkg/jsk_visualization/jsk_rqt_plugins [support_yn_btn_in_kinetic]
$
```

As shown above, `KeyboardInterrupt` message appears in spite of no keyboard interrupt in fact.
Also, `rostest` itself passed even though `required` tag is set to `true`.

So I think `required` tag doesn't work for me and created this PR.

**Comment by YutoUchimi** [comment]
melodic test is failing because of server error...? (HTTP status code: 503)

```
++ curl https://bootstrap.pypa.io/get-pip.py

++ sudo python -

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                                 Dload  Upload   Total   Spent    Left  Speed


  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   108    0   108    0     0    226      0 --:--:-- --:--:-- --:--:--   226

  File "<stdin>", line 1

    <html><body><h1>503 Service Unavailable</h1>

    ^

SyntaxError: invalid syntax

+++ error

+++ travis_time_end 31

+++ set +x
```

**Comment by YutoUchimi** [comment]
It seems that `openni_launch` is not released in melodic, and that causes failure in testing `pr2_play.launch`

```
+ rosdep install -q -y --rosdistro melodic -n -q -r --ignore-src --from-paths /opt/ros/melodic/share .

[33mYou are using pip version 9.0.3, however version 18.0 is available.

You should consider upgrading via the 'pip install --upgrade pip' command.[0m

[33mYou are using pip version 9.0.3, however version 18.0 is available.

You should consider upgrading via the 'pip install --upgrade pip' command.[0m

[33mYou are using pip version 9.0.3, however version 18.0 is available.

You should consider upgrading via the 'pip install --upgrade pip' command.[0m

ERROR: the following packages/stacks could not have their rosdep keys resolved

to system dependencies:

jsk_tilt_laser: Cannot locate rosdep definition for [multisense_lib]

jsk_tools: Cannot locate rosdep definition for [pr2_computer_monitor]

jsk_data: Cannot locate rosdep definition for [openni_launch]

Continuing to install resolvable dependencies...
```
```
[jsk_data:make] [35m[1mScanning dependencies of target _run_tests_jsk_data_roslaunch-check_launch_pr2_play.launch[0m




                                                                                
[jsk_data:make] make[3]: Leaving directory '/home/travis/ros/ws_jsk_common/build/jsk_data'




                                                                                
[jsk_data:make] /usr/bin/make -f CMakeFiles/_run_tests_jsk_data_roslaunch-check_launch_pr2_play.launch.dir/build.make CMakeFiles/_run_tests_jsk_data_roslaunch-check_launch_pr2_play.launch.dir/build




                                                                                
[jsk_data:make] make[3]: Entering directory '/home/travis/ros/ws_jsk_common/build/jsk_data'



                                                                                
[jsk_data:make] catkin_generated/env_cached.sh /usr/bin/python2 /opt/ros/melodic/share/catkin/cmake/test/run_tests.py /home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data/roslaunch-check_launch_pr2_play.launch.xml "/usr/bin/cmake -E make_directory /home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data" "/opt/ros/melodic/share/roslaunch/cmake/../scripts/roslaunch-check -o '/home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data/roslaunch-check_launch_pr2_play.launch.xml' '/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_data/launch/pr2_play.launch' "




                                                  
                                                                                
[jsk_data:make] [93mxacro.py is deprecated; please use xacro instead[0m




                                                                                
[jsk_data:make] Traceback (most recent call last):



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/share/roslaunch/cmake/../scripts/roslaunch-check", line 90, in <module>



                                                                                
[jsk_data:make]     error_msg = check_roslaunch_file(roslaunch_path, use_test_depends=options.test_depends)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/share/roslaunch/cmake/../scripts/roslaunch-check", line 47, in check_roslaunch_file



                                                                                
[jsk_data:make]     error_msg = roslaunch.rlutil.check_roslaunch(roslaunch_file, use_test_depends=use_test_depends)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/rlutil.py", line 193, in check_roslaunch



                                                                                
[jsk_data:make]     rl_config = roslaunch.config.load_config_default([f], DEFAULT_MASTER_PORT, verbose=False)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/config.py", line 457, in load_config_default




                                                                                
[jsk_data:make]     loader.load(f, config, argv=args, verbose=verbose)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 751, in load



                                                                                
[jsk_data:make]     self._load_launch(launch, ros_config, is_core=core, filename=filename, argv=argv, verbose=verbose)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 723, in _load_launch



                                                                                
[jsk_data:make]     self._recurse_load(ros_config, launch.childNodes, self.root_context, None, is_core, verbose)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 687, in _recurse_load




                                                                                
[jsk_data:make]     val = self._include_tag(tag, context, ros_config, default_machine, is_core, verbose)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 95, in call



                                                                                
[jsk_data:make]     return f(*args, **kwds)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 591, in _include_tag




                                                                                
[jsk_data:make]     inc_filename = self.resolve_args(tag.attributes['file'].value, context)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 185, in resolve_args



                                                                                
[jsk_data:make]     return substitution_args.resolve_args(args, context=context.resolve_dict, resolve_anon=self.resolve_anon)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 392, in resolve_args



                                                                                
[jsk_data:make]     resolved = _resolve_args(resolved, context, resolve_anon, commands)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 405, in _resolve_args




                                                                                
[jsk_data:make]     resolved = commands[command](resolved, a, args, context)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 166, in _find



                                                                                
[jsk_data:make]     source_path_to_packages=source_path_to_packages)



                                                                                
[jsk_data:make]   File "/opt/ros/melodic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 212, in _find_executable



                                                                                
[jsk_data:make]     full_path = _get_executable_path(rp.get_path(args[0]), path)



                                                                                
[jsk_data:make]   File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 203, in get_path




                                                                                
[jsk_data:make]     raise ResourceNotFound(name, ros_paths=self._ros_paths)



                                                                                
[jsk_data:make] rospkg.common.ResourceNotFound: openni_launch



                                                                                
[jsk_data:make] ROS path [0]=/opt/ros/melodic/share/ros



                                                                                
[jsk_data:make] ROS path [1]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_data



                                                                                
[jsk_data:make] ROS path [2]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_coordination_system/dynamic_tf_publisher



                                                                                
[jsk_data:make] ROS path [3]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_ros_patch/image_view2



                                                                                
[jsk_data:make] ROS path [4]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_data



                                                                                
[jsk_data:make] ROS path [5]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_network_tools



                                                                                
[jsk_data:make] ROS path [6]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_tilt_laser



                                                                                
[jsk_data:make] ROS path [7]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_tools




                                                                                
[jsk_data:make] ROS path [8]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_topic_tools



                                                                                
[jsk_data:make] ROS path [9]=/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_ros_patch/multi_map_server



                                                                                
[jsk_data:make] ROS path [10]=/home/travis/ros/ws_jsk_common/src/jsk_common/virtual_force_publisher



                                                                                
[jsk_data:make] ROS path [11]=/opt/ros/melodic/share




                                                                                
[jsk_data:make] Cannot find results, writing failure results to '/home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data/MISSING-roslaunch-check_launch_pr2_play.launch.xml'



                              
[jsk_data:make] -- run_tests.py: execute commands



                                                                                
[jsk_data:make]   /usr/bin/cmake -E make_directory /home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data



                                                                                
[jsk_data:make]   /opt/ros/melodic/share/roslaunch/cmake/../scripts/roslaunch-check -o '/home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data/roslaunch-check_launch_pr2_play.launch.xml' '/home/travis/ros/ws_jsk_common/src/jsk_common/jsk_data/launch/pr2_play.launch' 




                                                                                
[jsk_data:make] checking /home/travis/ros/ws_jsk_common/src/jsk_common/jsk_data/launch/pr2_play.launch




                                                                                
[jsk_data:make] -- run_tests.py: verify result "/home/travis/ros/ws_jsk_common/build/jsk_data/test_results/jsk_data/roslaunch-check_launch_pr2_play.launch.xml"




                                                                                
[jsk_data:make] make[3]: Leaving directory '/home/travis/ros/ws_jsk_common/build/jsk_data'




                                                                                
[jsk_data:make] Built target _run_tests_jsk_data_roslaunch-check_launch_pr2_play.launch
```

**Comment by k-okada** [comment]
> So I think required tag doesn't work for me and created this PR.

can you increase the timeout for test node? because Keyboard Interrupt means rostest/roslaunch teminates the node, not the node terminates because of error.

**Comment by YutoUchimi** [comment]
Added `import hoge` to 3rd line of `plot_2d.py`, which is imported from `rqt_2d_plot`.
And set `time-limit` to 3600, `retry` to 3 in test_always_pass.launch.

The result is:
```
$ rostest jsk_rqt_plugins test_always_pass.test 
... logging to /home/yutouchimi/.ros/log/rostest-libra-6997.log
[ROSUNIT] Outputting test results to /home/yutouchimi/.ros/test_results/jsk_rqt_plugins/rostest-test_test_always_pass.xml
Traceback (most recent call last):
  File "/home/yutouchimi/Projects/mirror_recognition/src/jsk-ros-pkg/jsk_visualization/jsk_rqt_plugins/bin/rqt_2d_plot", line 9, in <module>
    from jsk_rqt_plugins.plot_2d import Plot2D
  File "/home/yutouchimi/Projects/mirror_recognition/src/jsk-ros-pkg/jsk_visualization/jsk_rqt_plugins/src/jsk_rqt_plugins/plot_2d.py", line 3, in <module>
    import hoge
ImportError: No module named hoge
[Testcase: testtest_always_pass] ... ok

[ROSTEST]-----------------------------------------------------------------------

[jsk_rqt_plugins.rosunit-test_always_pass/test_always_pass][passed]

SUMMARY
 * RESULT: SUCCESS
 * TESTS: 1
 * ERRORS: 0
 * FAILURES: 0

rostest log file is in /home/yutouchimi/.ros/log/rostest-libra-6997.log
```

This time `ImportError` appears as we expected, but `rostest` results in success...

**Comment by k-okada** [comment]
oh, i am wrong, set duration 1 sec.
```

Scanning dependencies of target _run_tests_hoge_rostest_hoge.test
make[3]: Leaving directory '/tmp/hoge/build/hoge'
/usr/bin/make -f
CMakeFiles/_run_tests_hoge_rostest_hoge.test.dir/build.make
CMakeFiles/_run_tests_hoge_rostest_hoge.test.dir/build
make[3]: Entering directory '/tmp/hoge/build/hoge'
catkin_generated/env_cached.sh /usr/bin/python
/opt/ros/kinetic/share/catkin/cmake/test/run_tests.py
/tmp/hoge/build/hoge/test_results/hoge/rostest-hoge.xml
/opt/ros/kinetic/share/rostest/cmake/../../../bin/rostest\
--pkgdir=/tmp/hoge/src/hoge\ --package=hoge\ --results-filename\ hoge.xml\
--results-base-dir\ "/tmp/hoge/build/hoge/test_results"\
/tmp/hoge/src/hoge/hoge.test\
-- run_tests.py: execute commands
  /opt/ros/kinetic/share/rostest/cmake/../../../bin/rostest
--pkgdir=/tmp/hoge/src/hoge --package=hoge --results-filename hoge.xml
--results-base-dir /tmp/hoge/build/hoge/test_results
/tmp/hoge/src/hoge/hoge.test
... logging to /home/k-okada/.ros/log/rostest-p51s-23441.log
[ROSUNIT] Outputting test results to
/tmp/hoge/build/hoge/test_results/hoge/rostest-hoge.xml
;; loading roseus("1.7.1") on euslisp((9.25 ip-172-30-1-163 Wed Jul 18
21:49:04 PST 2018  1.2.0))
Call Stack (max depth: 20):
/opt/ros/kinetic/share/euslisp/jskeus/eus/Linux64/bin/irteusgl
roseus-error: undefined function a in #<compiled-code #X58317e8>,
exitting...
Traceback (most recent call last):
  File "/opt/ros/kinetic/share/rostest/nodes/hztest", line 245, in <module>
    time.sleep(0.75)
KeyboardInterrupt
[Testcase: testalways_true] ... FAILURE!
FAILURE: test [always_true] did not generate test results
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/runner.py",
line 166, in fn
    self.assert_(os.path.isfile(test_file), "test [%s] did not generate
test results"%test_name)
  File "/usr/lib/python2.7/unittest/case.py", line 422, in assertTrue
    raise self.failureException(msg)
--------------------------------------------------------------------------------

results===============
['__class__', '__delattr__', '__doc__', '__format__', '__getattribute__',
'__hash__', '__init__', '__module__', '__new__', '__reduce__',
'__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__',
'__str__', '__subclasshook__', 'accumulate', 'add_test_case_result',
'name', 'num_errors', 'num_failures', 'num_tests', 'system_err',
'system_out', 'test_case_results', 'time', 'xml_element']
[ROSTEST]-----------------------------------------------------------------------

[testalways_true][failed]

SUMMARY
 * RESULT: FAIL
 * TESTS: 0
 * ERRORS: 0
 * FAILURES: 1

ERROR: The following tests failed to run:
 * testalways_true

rostest log file is in /home/k-okada/.ros/log/rostest-p51s-23441.log
-- run_tests.py: verify result
"/tmp/hoge/build/hoge/test_results/hoge/rostest-hoge.xml"
make[3]: Leaving directory '/tmp/hoge/build/hoge'
Built target _run_tests_hoge_rostest_hoge.test
```

for


```
<launch>
  <node name="fail_node" pkg="roseus" type="roseus" args="(a)"
required="true" />

  <!-- to pass always_true" -->
  <node name="pub" pkg="roscpp_tutorials" type="talker" />

  <param name="hztest1/topic" value="chatter" />
  <param name="hztest1/hz" value="10.0" />
  <param name="hztest1/hzerror" value="10" />
  <param name="hztest1/test_duration" value="5.0" />
  <test test-name="always_true" pkg="rostest" type="hztest" name="hztest1"
/>
</launch>
```



--
◉ Kei Okada



2018年9月13日(木) 20:11 Yuto Uchimi <notifications@github.com>:

**Comment by YutoUchimi** [comment]
Hmm... `required="true"` causes failure in rostest and is useful because `time.sleep(0.75)` in `hztest` waits for other nodes starting before running test.
```
$ cat $(rospack find rostest)/nodes/hztest | tail -n 12
if __name__ == '__main__':
    # A dirty hack to work around an apparent race condition at startup
    # that causes some hztests to fail.  Most evident in the tests of
    # rosstage.
    time.sleep(0.75)
    try:
        rostest.run('rostest', NAME, HzTest, sys.argv)
    except KeyboardInterrupt:
        pass
    print("exiting")

        
yutouchimi@libra:~
```

So, I can test if some nodes raise `ImportError` without this PR (as @k-okada advised me).

Note that this [`required` attribute + `hztest`] method can be not so usefull when we want to check nodes launched from `include` tag, because, needless to say, `include` tag has no attribute named `required`.

**Comment by YutoUchimi** [comment]
Shall I close this PR or put this PR off for a while?

**Pull request closed without merging**